src/linear_algebra/matrix/to_linear_equiv.lean

```diff
@@ -92,4 +97,86 @@ range_eq_top.mpr (to_linear_equiv b A hA).surjective
+      exact (linear_equiv.of_injective_endo M.to_lin' this).apply_symm_apply v },
+    exact matrix.det_ne_zero_of_right_inverse this }
+end
+
+lemma exists_mul_vec_eq_zero_iff {A : Type*} [decidable_eq n] [integral_domain A]
+  {M : matrix n n A} :
+  (∃ (v ≠ 0), M.mul_vec v = 0) ↔ M.det = 0 :=
+begin
+  have : (∃ (v ≠ 0), mul_vec ((algebra_map A (fraction_ring A)).map_matrix M) v = 0) ↔ _ :=
+    exists_mul_vec_eq_zero_iff_aux,
+  rw [← ring_hom.map_det, is_fraction_ring.to_map_eq_zero_iff] at this,
+  refine iff.trans _ this, split; rintro ⟨v, hv, mul_eq⟩,
+  { refine ⟨λ i, algebra_map _ _ (v i), mt (λ h, funext $ λ i, _) hv, _⟩,
+    { exact is_fraction_ring.injective A (fraction_ring A) (congr_fun h i) },
+    { ext i,
+      refine (ring_hom.map_mul_vec _ _ _ i).symm.trans _,
+      rw [mul_eq, pi.zero_apply, ring_hom.map_zero, pi.zero_apply] } },
+  { letI := classical.dec_eq (fraction_ring A),
+    obtain ⟨⟨b, hb⟩, ba_eq⟩ := is_localization.exist_integer_multiples_of_finset
+      (non_zero_divisors A) (finset.univ.image v),
+    choose f hf using ba_eq,
+    refine ⟨λ i, f _ (finset.mem_image.mpr ⟨i, finset.mem_univ i, rfl⟩),
+            mt (λ h, funext $ λ i, _) hv, _⟩,
+    { have := congr_arg (algebra_map A (fraction_ring A)) (congr_fun h i),
+      rw [hf, subtype.coe_mk, pi.zero_apply, ring_hom.map_zero, algebra.smul_def,
+          mul_eq_zero, is_fraction_ring.to_map_eq_zero_iff] at this,
+      exact this.resolve_left (mem_non_zero_divisors_iff_ne_zero.mp hb), },
+    { ext i,
+      refine is_fraction_ring.injective A (fraction_ring A) _,
+      calc algebra_map A (fraction_ring A) (M.mul_vec (λ (i : n), f (v i) _) i)
+          = ((algebra_map A (fraction_ring A)).map_matrix M).mul_vec
+              (algebra_map _ (fraction_ring A) b • v) i : _
+      ... = 0 : _,
+      { simp_rw [ring_hom.map_mul_vec, mul_vec, dot_product, function.comp_app, hf,
+          subtype.coe_mk, ring_hom.map_matrix_apply, pi.smul_apply, smul_eq_mul,
+          algebra.smul_def] },
+      { rw [mul_vec_smul, mul_eq, pi.smul_apply, pi.zero_apply, smul_zero] } } },
+end
+
+lemma exists_vec_mul_eq_zero_iff {A : Type*} [decidable_eq n] [integral_domain A]
+  {M : matrix n n A} :
+  (∃ (v ≠ 0), M.vec_mul v = 0) ↔ M.det = 0 :=
+by simpa only [← M.det_transpose, ← mul_vec_transpose] using exists_mul_vec_eq_zero_iff
+
+theorem nondegenerate_iff_det_ne_zero {A : Type*} [decidable_eq n] [integral_domain A]
+  {M : matrix n n A} :
+  nondegenerate M ↔ M.det ≠ 0 :=
+begin
+  refine iff.trans _ (not_iff_not.mpr exists_vec_mul_eq_zero_iff),
+  simp only [not_exists],
+  split,
+  { intros hM v hv hMv,
+    obtain ⟨w, hwMv⟩ := hM.exists_not_ortho_of_ne_zero hv,
+    simpa only [dot_product_mul_vec, hMv, zero_dot_product] using hwMv },
+  { intros h v hv,
+    refine not_imp_not.mp (h v) (funext $ λ i, _),
+    simpa only [dot_product_mul_vec, dot_product_single, mul_one] using hv (pi.single i 1) }
+end
+
```

Comment:
Could you please add `nondegenerate.det_ne_zero` for easy use with dot notation?